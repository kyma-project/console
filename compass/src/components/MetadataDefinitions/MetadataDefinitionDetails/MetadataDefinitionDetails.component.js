import React, { useState } from 'react';
import { ActionBar } from 'fundamental-react';
import {
  Button,
  Breadcrumb,
  Panel,
  PanelHead,
  PanelHeader,
  PanelBody,
} from '@kyma-project/react-components';
import LuigiClient from '@kyma-project/luigi-client';

import '../../../shared/styles/header.scss';
import ResourceNotFound from '../../Shared/ResourceNotFound.component';
import JSONEditorComponent from '../../Shared/JSONEditor';

const MetadataDefinitionDetails = ({
  metadataDefinition: metadataDefinitionQuery,
  updateLabelDefinition,
}) => {
  const [isSchemaValid, setSchemaValid] = useState(true);
  const [schema, setSchema] = useState({});

  const handleSchemaChange = schema => {
    try {
      setSchema(JSON.parse(schema));

      setSchemaValid(true);
    } catch {
      setSchemaValid(false);
    }
  };

  const handleSaveChanges = async definitionKey => {
    try {
      await updateLabelDefinition({
        key: definitionKey,
        schema,
      });
      metadataDefinitionQuery.refetch(); //  to format the JSON
      // onCompleted(runtimeName, `Runtime created succesfully`);
    } catch (e) {
      //onError(`The runtime could not be created succesfully`, e.message || ``);
    }
  };

  const metadataDefinition =
    (metadataDefinitionQuery && metadataDefinitionQuery.labelDefinition) || {};
  const loading = metadataDefinitionQuery.loading;
  const error = metadataDefinitionQuery.error;

  if (!metadataDefinitionQuery) {
    if (loading) return 'Loading...';
    if (error)
      return (
        <ResourceNotFound
          resource="Metadata Definition"
          breadcrumb="MetadataDefinitions"
        />
      );
    return null;
  }
  if (error) {
    return `Error! ${error.message}`;
  }

  return (
    <>
      <header className="fd-has-background-color-background-2">
        <section className="fd-has-padding-regular fd-has-padding-bottom-none action-bar-wrapper">
          <section>
            <Breadcrumb>
              <Breadcrumb.Item
                name="Metadata Definitions"
                url="#"
                onClick={() =>
                  LuigiClient.linkManager()
                    .fromClosestContext()
                    .navigate(`/metadata-definitions`)
                }
              />
              <Breadcrumb.Item />
            </Breadcrumb>
            <ActionBar.Header
              title={metadataDefinition.key || 'Loading name...'}
            />
          </section>
          <ActionBar.Actions>
            <Button
              onClick={() => handleSaveChanges(metadataDefinition.key)}
              disabled={!isSchemaValid}
              option="emphasized"
            >
              Save
            </Button>
          </ActionBar.Actions>
        </section>
      </header>
      <section className="fd-section">
        <Panel>
          <PanelHeader>
            <PanelHead title="Schema" />
          </PanelHeader>

          <PanelBody>
            <JSONEditorComponent
              onChangeText={handleSchemaChange}
              text={JSON.stringify(metadataDefinition.schema || {}, null, 2)}
            />
          </PanelBody>
        </Panel>
      </section>
    </>
  );
};

export default MetadataDefinitionDetails;
