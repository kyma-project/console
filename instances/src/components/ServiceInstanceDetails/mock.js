export const Mock = {
  __typename: 'clusterDocsTopic',
  name: 'kyma',
  groupName: 'root',
  displayName: 'Kyma',
  description: 'Overall documentation for Kyma',
  assets: [
    {
      __typename: 'clusterAsset',
      name: 'asset1',
      type: 'markdown',
      files: [
        {
          __typename: 'file',
          url:
            'https://raw.githubusercontent.com/kyma-project/kyma/master/docs/kyma/docs/01-01-in-a-nutshell.md',
          metadata: {
            __typename: 'fileMetadata',
            title: 'In a nutshell',
            type: 'Overview',
          },
        },
        {
          __typename: 'file',
          url:
            'https://raw.githubusercontent.com/kyma-project/kyma/master/docs/kyma/docs/04-01-overview.md',
          metadata: {
            __typename: 'fileMetadata',
            title: 'Overview',
            type: 'Installation',
          },
        },
        {
          __typename: 'file',
          url:
            'https://raw.githubusercontent.com/kyma-project/kyma/master/docs/kyma/docs/04-03-cluster-installation.md',
          metadata: {
            __typename: 'fileMetadata',
            // title: 'Install Kyma on a cluster',
            type: 'Installation',
          },
        },
        {
          __typename: 'file',
          url:
            'https://raw.githubusercontent.com/kyma-project/kyma/master/docs/kyma/docs/12-01-try-out-kyma.md',
          metadata: {
            __typename: 'fileMetadata',
            type: 'Examples',
            title: 'Try out Kyma',
          },
        },
      ],
    },
  ],
};
