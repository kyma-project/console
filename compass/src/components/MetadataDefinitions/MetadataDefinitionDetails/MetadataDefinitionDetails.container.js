import { graphql, compose } from 'react-apollo';

import {
  DELETE_LABEL_DEFINITION,
  GET_LABEL_DEFINITION,
  UPDATE_LABEL_DEFINITION,
} from '../gql';
import { SEND_NOTIFICATION } from '../../../gql';

import MetadataDefinitionDetails from './MetadataDefinitionDetails.component';

export default compose(
  graphql(GET_LABEL_DEFINITION, {
    name: 'metadataDefinition',
    options: props => {
      return {
        fetchPolicy: 'cache-and-network',
        errorPolicy: 'all',
        variables: {
          key: props.definitionKey,
        },
      };
    },
  }),
  graphql(UPDATE_LABEL_DEFINITION, {
    props: ({ mutate, error }) => ({
      updateLabelDefinition: data =>
        mutate({
          variables: {
            in: data,
          },
        }),
    }),
  }),
  graphql(DELETE_LABEL_DEFINITION, {
    props: ({ mutate }) => ({
      deleteLabelDefinition: key =>
        mutate({
          variables: {
            key: key,
          },
        }),
    }),
  }),
  graphql(SEND_NOTIFICATION, {
    name: 'sendNotification',
  }),
)(MetadataDefinitionDetails);
