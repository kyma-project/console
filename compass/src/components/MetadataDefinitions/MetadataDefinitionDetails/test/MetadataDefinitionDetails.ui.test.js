import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import { mount } from 'enzyme';
import { mocks } from './mock';
import { ActionBar } from 'fundamental-react';
import { Panel } from '@kyma-project/react-components';
import MetadataDefinitionDetails from '../MetadataDefinitionDetails.container';
import JSONEditorComponent from '../../../Shared/JSONEditor';

describe('MetadataDefinitionDetails UI', () => {
  console.error = jest.fn();
  console.warn = jest.fn();
  afterEach(() => {
    console.error.mockReset();
    console.warn.mockReset();
  });

  it(`Renders "Loading name..." when there's no GQL response`, async () => {
    const component = mount(
      <MockedProvider addTypename={false}>
        <MetadataDefinitionDetails definitionKey="noschemalabel" />
      </MockedProvider>,
    );

    await wait(0); // wait for response

    expect(component.find(ActionBar.Header).text()).toEqual('Loading name...');
  });

  it(`Renders the name `, async () => {
    const component = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MetadataDefinitionDetails definitionKey="noschemalabel" />
      </MockedProvider>,
    );

    await wait(0); // wait for response

    component.update();
    expect(component.find(ActionBar.Header).text()).toEqual(
      mocks[0].result.data.labelDefinition.key,
    );
    expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
  });

  it(`Clicking "Save" triggers the mutation`, async () => {
    const component = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MetadataDefinitionDetails definitionKey="noschemalabel" />
      </MockedProvider>,
    );

    await wait(0); // wait for response
    expect(mocks[2].result.mock.calls.length).toEqual(0);

    component.find('.fd-button--emphasized').simulate('click');

    await wait(0); // wait for response

    expect(mocks[2].result.mock.calls.length).toEqual(1);

    expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    expect(console.warn.mock.calls[0][0]).toMatchSnapshot(); // Apollo's @client warning because of Notification
  });

  describe('The schema is provided', () => {
    it(`Renders panel with toggle set to on `, async () => {
      const component = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
          <MetadataDefinitionDetails definitionKey="labelWithInvalidSchema" />
        </MockedProvider>,
      );

      await wait(0); // wait for response

      component.update();

      expect(
        component
          .find(Panel)
          .find('Toggle')
          .prop('checked'),
      ).toEqual(true);
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });

    it(`Renders JSON editor`, async () => {
      const component = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
          <MetadataDefinitionDetails definitionKey="labelWithInvalidSchema" />
        </MockedProvider>,
      );

      await wait(0); // wait for response

      component.update();

      expect(component.find(Panel).exists(JSONEditorComponent)).toBeTruthy();
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });
  });
  describe('The schema is not provided', () => {
    it(`Renders panel with toggle set to off`, async () => {
      const component = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
          <MetadataDefinitionDetails definitionKey="noschemalabel" />
        </MockedProvider>,
      );

      await wait(0); // wait for response

      component.update();

      expect(
        component
          .find(Panel)
          .find('Toggle')
          .prop('checked'),
      ).toBeUndefined();
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });

    it(`Doesn't render JSON editor`, async () => {
      const component = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
          <MetadataDefinitionDetails definitionKey="noschemalabel" />
        </MockedProvider>,
      );

      await wait(0); // wait for response

      component.update();

      expect(
        component.find(Panel).exists(JSONEditorComponent),
      ).not.toBeTruthy();
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });
  });
});
