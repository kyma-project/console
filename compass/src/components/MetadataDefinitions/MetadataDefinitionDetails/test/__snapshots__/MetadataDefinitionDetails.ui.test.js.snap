// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataDefinitionDetails UI Renders the name  1`] = `"Warning: Each child in a list should have a unique \\"key\\" prop. See https://fb.me/react-warning-keys for more information.%s"`;

exports[`MetadataDefinitionDetails UI The schema is not provided Renders panel with toggle set to off 1`] = `"Warning: Each child in a list should have a unique \\"key\\" prop. See https://fb.me/react-warning-keys for more information.%s"`;
